display:none HTMLMediaElement with a camera track as source is leaking video frames

youennf opened the pull request:
#### 3d75fe46264aa2aaa58b3d85eeece0d53d8c821b
<pre>
display:none HTMLMediaElement with a camera track as source is leaking video frames
<a href="https://bugs.webkit.org/show_bug.cgi?id=242939">https://bugs.webkit.org/show_bug.cgi?id=242939</a>
rdar://96677785

Reviewed by Eric Carlson.

AVSampleBufferDisplayLayer is keeping CVPixelBuffers when it cannot display them, which happens on iOS in case HTMLMediaElement is display:none.
In that case, the camera buffer pool might get exhausted and this will fail camera capture.
We fixed this by making sure to not enqueue frames in that case, but this got broken in <a href="https://github.com/WebKit/WebKit/commit/d126b89f7caa23c65c4d1df42312d2e6bcf74725.">https://github.com/WebKit/WebKit/commit/d126b89f7caa23c65c4d1df42312d2e6bcf74725.</a>
We reintroduce a specific callback renderVideoWillBeDestroyed so that only MediaPlayerPrivateMediaStreamAVFObjC is changed.

To make our mock camera closer to actual cameras, we create CVPixelBuffer from a pool with kCVPixelBufferPoolAllocationThresholdKey.
This should allow to test CVPixelBuffer camera leaks.

Update LayoutTests/fast/mediastream/resources/getUserMedia-to-canvas.js to make sure we do not keep a VideoFrame for each video element as otherwise, we can run out of camera frames.

* LayoutTests/fast/mediastream/getUserMedia-media-element-display-none-expected.txt: Added.
* LayoutTests/fast/mediastream/getUserMedia-media-element-display-none.html: Added.
* LayoutTests/fast/mediastream/resources/getUserMedia-to-canvas.js:
(createUserMediaToCanvasTests):
* Source/WebCore/platform/cocoa/CoreVideoSoftLink.cpp:
* Source/WebCore/platform/cocoa/CoreVideoSoftLink.h:
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::renderVideoWillBeDestroyed):
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::renderVideoWillBeDestroyed):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.h:
* Source/WebCore/platform/graphics/cv/CVUtilities.h:
* Source/WebCore/platform/graphics/cv/CVUtilities.mm:
(WebCore::createCVPixelBufferFromPool):
* Source/WebCore/platform/graphics/cv/ImageTransferSessionVT.h:
(WebCore::ImageTransferSessionVT::setMaximumBufferPoolSize):
* Source/WebCore/platform/graphics/cv/ImageTransferSessionVT.mm:
(WebCore::ImageTransferSessionVT::convertPixelBuffer):
* Source/WebCore/platform/mediastream/mac/MockRealtimeVideoSourceMac.h:
* Source/WebCore/platform/mediastream/mac/MockRealtimeVideoSourceMac.mm:
(WebCore::MockRealtimeVideoSourceMac::updateSampleBuffer):
* Source/WebCore/rendering/RenderVideo.cpp:
(WebCore::RenderVideo::willBeDestroyed):

Canonical link: <a href="https://commits.webkit.org/252871@main">https://commits.webkit.org/252871@main</a>
</pre>

jernoble commented on Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
```diff
@@ -336,6 +336,8 @@ class MediaPlayerPrivateInterface {
 
     virtual String errorMessage() const { return { }; }
+
+    virtual void renderVideoWillBeDestroyed() { }
```
I don't understand why `acceleratedRenderingStateChanged()` is insufficient here. Destroying the renderer should result in that method being called, and `renderingCanBeAccelerated()` should start returning `false`, and `MediaPlayerPrivateMediaStreamAVFObjC::destroyLayers()` should be called.

jernoble replied:
Ah, but that would affect all MediaPlayerPrivateInterface subclasses, and you're trying to do the less risky change here.

youennf replied:
Right, I am only changing MediaPlayerPrivateMediaStreamAVFObjC in this patch.

eric-carlson reviewed (approved):
r=me

youennf committed:
display:none HTMLMediaElement with a camera track as source is leaking video frames
https://bugs.webkit.org/show_bug.cgi?id=242939
rdar://96677785

Reviewed by Eric Carlson.

AVSampleBufferDisplayLayer is keeping CVPixelBuffers when it cannot display them, which happens on iOS in case HTMLMediaElement is display:none.
In that case, the camera buffer pool might get exhausted and this will fail camera capture.
We fixed this by making sure to not enqueue frames in that case, but this got broken in https://github.com/WebKit/WebKit/commit/d126b89f7caa23c65c4d1df42312d2e6bcf74725.
We reintroduce a specific callback renderVideoWillBeDestroyed so that only MediaPlayerPrivateMediaStreamAVFObjC is changed.

To make our mock camera closer to actual cameras, we create CVPixelBuffer from a pool with kCVPixelBufferPoolAllocationThresholdKey.
This should allow to test CVPixelBuffer camera leaks.

Update LayoutTests/fast/mediastream/resources/getUserMedia-to-canvas.js to make sure we do not keep a VideoFrame for each video element as otherwise, we can run out of camera frames.

* LayoutTests/fast/mediastream/getUserMedia-media-element-display-none-expected.txt: Added.
* LayoutTests/fast/mediastream/getUserMedia-media-element-display-none.html: Added.
* LayoutTests/fast/mediastream/resources/getUserMedia-to-canvas.js:
(createUserMediaToCanvasTests):
* Source/WebCore/platform/cocoa/CoreVideoSoftLink.cpp:
* Source/WebCore/platform/cocoa/CoreVideoSoftLink.h:
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::renderVideoWillBeDestroyed):
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::renderVideoWillBeDestroyed):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.h:
* Source/WebCore/platform/graphics/cv/CVUtilities.h:
* Source/WebCore/platform/graphics/cv/CVUtilities.mm:
(WebCore::createCVPixelBufferFromPool):
* Source/WebCore/platform/graphics/cv/ImageTransferSessionVT.h:
(WebCore::ImageTransferSessionVT::setMaximumBufferPoolSize):
* Source/WebCore/platform/graphics/cv/ImageTransferSessionVT.mm:
(WebCore::ImageTransferSessionVT::convertPixelBuffer):
* Source/WebCore/platform/mediastream/mac/MockRealtimeVideoSourceMac.h:
* Source/WebCore/platform/mediastream/mac/MockRealtimeVideoSourceMac.mm:
(WebCore::MockRealtimeVideoSourceMac::updateSampleBuffer):
* Source/WebCore/rendering/RenderVideo.cpp:
(WebCore::RenderVideo::willBeDestroyed):

Canonical link: https://commits.webkit.org/252871@main
